src/Features/CSharp/Portable/CodeFixes/GenerateMethod/GenerateDeconstructMethodCodeFixProvider.cs

```diff
@@ -0,0 +1,93 @@
+    internal class GenerateDeconstructMethodCodeFixProvider : CodeFixProvider
+#pragma warning restore RS1016 // Code fix providers should provide FixAll support.
+    {
+        private const string CS8129 = nameof(CS8129); // No suitable Deconstruct instance or extension method was found...
+
+        public sealed override ImmutableArray<string> FixableDiagnosticIds
+        {
+            get { return ImmutableArray.Create(CS8129); }
+        }
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            // Not supported in REPL
+            if (context.Project.IsSubmission)
+            {
+                return;
+            }
+
+            var document = context.Document;
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var span = context.Span;
+            var token = root.FindToken(span.Start);
+
+            var deconstruction = token.GetAncestors<SyntaxNode>()
+                .FirstOrDefault(n => n.IsKind(SyntaxKind.SimpleAssignmentExpression, SyntaxKind.ForEachVariableStatement));
+
+            if (deconstruction is null)
+            {
+                Debug.Assert(false);
+                return;
+            }
+
+            var model = await document.GetSemanticModelForNodeAsync(deconstruction, context.CancellationToken).ConfigureAwait(false);
+
+            DeconstructionInfo info;
+            ITypeSymbol type;
+            ExpressionSyntax target;
+            switch (deconstruction)
+            {
+                case ForEachVariableStatementSyntax @foreach:
+                    info = model.GetDeconstructionInfo(@foreach);
+                    type = model.GetForEachStatementInfo(@foreach).ElementType;
+                    target = @foreach.Variable;
+                    break;
+                case AssignmentExpressionSyntax assignment:
+                    info = model.GetDeconstructionInfo(assignment);
+                    type = model.GetTypeInfo(assignment.Right).Type;
+                    target = assignment.Left;
+                    break;
+                default:
+                    Debug.Assert(false);
+                    return;
+            }
+
+            if (type.Kind != SymbolKind.NamedType)
+            {
+                return;
+            }
+
+            if (info.Method != null || !info.Nested.IsEmpty)
+            {
+                // There is already a Deconstruct method, or we have a nesting situation
+                return;
+            }
+
+            var service = document.GetLanguageService<IGenerateDeconstructMemberService>();
+            var codeActions = await service.GenerateDeconstructMethodAsync(document, target, (INamedTypeSymbol)type, context.CancellationToken).ConfigureAwait(false);
+
+            if (!codeActions.IsDefaultOrEmpty)
+            {
+                context.RegisterFixes(codeActions, context.Diagnostics);
```

Comment:
nit.  if you make it so that GenerateDeconstructMethodAsync always returns a non-default array, you can just call context.RegisterFixes safely.